src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,3349 @@
+/*******************************************************************\
+
+Module: Generates string constraints to link results from string functions
+        with their arguments. This is inspired by the PASS paper at HVC'13:
+        "PASS: String Solving with Parameterized Array and Interval Automaton"
+        by Guodong Li and Indradeep Ghosh, which gives examples of constraints
+        for several functions.
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#include <ansi-c/string_constant.h>
+#include <java_bytecode/java_types.h>
+#include <solvers/floatbv/float_bv.h>
+#include <solvers/refinement/string_constraint_generator.h>
+#include <util/arith_tools.h>
+#include <util/pointer_predicates.h>
+#include <util/prefix.h>
+#include <util/ssa_expr.h>
+#include <util/unicode.h>
+
+/*******************************************************************\
+
+Function: string_constraint_generatort::constant_char
+
+  Inputs: integer representing a character, we do not use char type here
+          because java characters use more than 8 bits.
+
+ Outputs: constant expression corresponding to the character.
+
+ Purpose: generate constant character expression with character type.
+
+\*******************************************************************/
+
+constant_exprt string_constraint_generatort::constant_char(int i) const
+{
+  return from_integer(i, get_char_type());
+}
+
+typet string_constraint_generatort::get_char_type() const
+{
+  if(mode==ID_C)
+    return refined_string_typet::char_type();
+  else if(mode==ID_java)
+    return refined_string_typet::java_char_type();
+  else
+    assert(false); // only C and java modes supported
+}
+
+/*******************************************************************\
+
+Function: string_constraint_generatort::fresh_univ_index
+
+  Inputs: a prefix
+
+ Outputs: a symbol of index type whose name starts with the prefix
+
+ Purpose: generate an index symbol to be used as an universaly quantified
+          variable
+
+\*******************************************************************/
+
+symbol_exprt string_constraint_generatort::fresh_univ_index(
+  const irep_idt &prefix)
+{
+  return string_exprt::fresh_symbol(prefix, get_index_type());
```

Comment:
symbols generated by `string_exprt::fresh_symbol` start with `string_refinement#` and `#` shouldn't be allowed in the source (at least in java)